### PR TITLE
Prepare train.py for model chunks for pipelining

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #411
* #358
* __->__ #406

When using pipeline parallelism, a common technique  for reducing bubble
size is to use schedules that specify more than one model chunk per
physical rank.  e.g. pp degree 4 could have 8 pipeline stages, and rank
0 could have stage 0 and stage 4.

To generalize this concept without forking too much code in train.py, I
make 'model_parts' a new container that either contains one model for
non-PP or simple PP cases, and contains multiple model parts for complex
PP cases.

In general, this is tractable becuase we treat each model part the same:
we create one optimizer per model part, and one lr scheduler per
optimizer.  We apply spmd and compile individually to each model part.
The general pattern is to loop over the model parts and perform an
action on each part, which also works fine if the list size is 1.

The rest of train.py and optimizer/lr_scheduler changes add syntax sugar
to simplify calling a method on each model part or optimizer part.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -195,16 +216,21 @@ def loss_fn(pred, labels):
         logger.info(
             f"Building {model_name} {job_config.model.flavor} with {model_config}"
         )
-        model = model_cls.from_model_args(model_config)
+        whole_model = model_cls.from_model_args(model_config)
+
+    # In 1D/2D cases or PP with simple schedules, model_parts is just one item
+    # for PP with looped schedules, each item is one stage-model-chunk
+    # we iterate all model_parts for applying SPMD parallelism, compilation, optimizer, and checkpointing
+    model_parts = [whole_model]
```
Comment (tianyu-l) [inline train.py]: this can be moved to below until it's called for the first time

File: train.py [hunk 2]
```diff
@@ -219,26 +245,28 @@ def loss_fn(pred, labels):
...
-            model, world_mesh, parallel_dims, job_config, device, model_config
+        stages, model_parts = models_pipelining_fns[model_name](
+            whole_model, world_mesh, parallel_dims, job_config, device, model_config
         )
 
     # apply PT-D DP/TP parallelisms and activation checkpointing
-    model = models_parallelize_fns[model_name](
-        model, world_mesh, parallel_dims, job_config
+    model_parts = models_parallelize_fns[model_name](
+        model_parts, world_mesh, parallel_dims, job_config
```
Comment (tianyu-l) [inline train.py]: I wonder if it's better if we do the for loop on model_parts here, instead of in `parallelize_llama`. To me, it's just a simple for-loop so no need to hide the complexity.

File: torchtitan/checkpoint.py [hunk 3]
```diff
@@ -124,9 +124,9 @@ def checkpoint_mp(recv, send):
 class CheckpointManager:
     def __init__(
         self,
-        model: nn.Module,
-        optimizer: torch.optim.Optimizer,
-        lr_scheduler: torch.optim.lr_scheduler.LRScheduler,
+        model_parts: List[nn.Module],
+        optimizers: List[torch.optim.Optimizer],
+        lr_schedulers: List[torch.optim.lr_scheduler.LRScheduler],
```
Comment (tianyu-l) [inline torchtitan/checkpoint.py:129]: CheckpointManager and ModelWrapper/OptimizerWrapper have different styles -- one takes a module or a list of modules, the other always takes a list. Shall we make them consistent?

**Review by tianyu-l — COMMENTED** [review]
It seems these changes for PP are deeply coupling together with all parts of the scripts.

Instead of completely modifying these building blocks (e.g. `build_optimizer`), I wonder if it makes sense to create a class (call it `UtilsBuilder`, or `Trainer`, or something), which has methods `build_optimizer`, `build_lr_scheduler`, etc. Then PP just needs to create a subclass for it and be free to call same methods in `super`.

**Comment by tianyu-l** [comment]
another question: currently for PP do we do data loading on all the ranks? It seems we only need to do it on the first stage.

**Comment by wconstab** [comment]
> another question: currently for PP do we do data loading on all the ranks? It seems we only need to do it on the first stage.

good catch, it'd be good to avoid data loading on later PP stages. i might make the fix in a follow up PR if you don't mind, to keep things separate.

File: torchtitan/checkpoint.py (hunk 3 above)
Comment (wconstab) [reply torchtitan/checkpoint.py:129]: since i changed train.py to always have a list of model parts, we would never pass a single model inside anymore. i could update checkpoint to take a single model in ifs passed, but that codepath would not be exercised.

i could make the wrappers match this style, if you like that better?

File: torchtitan/utils.py [hunk 4]
```diff
@@ -17,6 +17,11 @@
 from torchtitan.parallelisms import ParallelDims
 
 
+def move_to_empty(model_parts: List[torch.nn.Module], device: torch.device):
+    for model in model_parts:
+        model.to_empty(device="cuda")
```
Comment (tianyu-l) [inline torchtitan/utils.py]: We probably should simply do this in `train.py` as well. It's similar to other for-loops in `train.py`, which are only called once in the code base.

**Comment by wconstab** [comment]
@tianyu-l I've cleaned up the simpler suggestions and still wondering what to do about the following.

> another question: currently for PP do we do data loading on all the ranks? It seems we only need to do it on the first stage.

For dataloading, I opened #411 - let's continue discussion on that PR

> CheckpointManager and ModelWrapper/OptimizerWrapper have different styles -- one takes a module or a list of modules, the other always takes a list. Shall we make them consistent?

LMK what you think about my question inline on this point.  Shall I change the wrappers?

> Instead of completely modifying these building blocks (e.g. build_optimizer), I wonder if it makes sense to create a class (call it UtilsBuilder, or Trainer, or something), which has methods build_optimizer, build_lr_scheduler, etc. Then PP just needs to create a subclass for it and be free to call same methods in super.

Do you think making these into a class/subclass is much different?  One thing I could do is move the optimizer builder into a same file as lr_scheduler builder.  (I'd have to do that anyway, if I made them into one class.  Or i could leave the 'build_optimizer' function outside of the 'build_optimizers' function instead of defining it inside.

File: torchtitan/checkpoint.py [hunk 5]
```diff
@@ -138,15 +138,28 @@ def __init__(
...
                 "dataloader": dataloader,
             }
         )
+        if len(lr_schedulers) == 1:
+            self.states["lr_scheduler"] = lr_schedulers[0]
+        else:
+            # For now, pipeline-parallel with looped schedules does not support resharding for lr_scheduler.
+            # It should only support saving and loading a distributed checkpoint with the same number of pp ranks
+            for idx, lr_scheduler in enumerate(lr_schedulers):
+                self.states[f"lr_scheduler_{idx}"] = lr_scheduler
```
Comment (fegin) [inline torchtitan/checkpoint.py:162]: I thought you are going to try to save only one copy of the `lr_scheduler` and assume all `lr_scheduler` have the same states.

File: train.py [hunk 6]
```diff
@@ -90,29 +90,49 @@ def load_state_dict(self, state_dict) -> None:
...
+            # TODO: make the optimizer options configurable by toml/cmd args
+            optimizer = torch.optim.Adam(model.parameters(), **optimizer_kwargs)
+        elif name == "AdamW":
+            optimizer = torch.optim.AdamW(model.parameters(), **optimizer_kwargs)
+        else:
+            raise NotImplementedError(f"Optimizer {name} not added.")
+
+        return optimizer
+
+    class OptimizersContainer:
```
Comment (fegin) [inline train.py:121]: Since we are going to have a class anyway, will it be better to have a `Optimizers` class to do everything, including the state_dict part?

```
class Optimizers:
    def __init__(self, model_parts, job_config: JobConfig) -> None:
            original build optimizer code
            state_dict init code
            
    def step(self):
             ...
             
     def  state_dict(self) -> None:
             OptimizerWrapper state_dict code
             
      def load_state_dict(self) -> None:
             OptimizerWrapper load_state_dict code
``` 

And can we also do this for model? It seems that the main training loop will need to loop all the model parts as well. Having a Models and a Optimizer classes can encapsulate the logic and remove extra classes (ModelWrapper and OptimizerWrapper).

**Review by tianyu-l — APPROVED** [review]
discussed offline with @wconstab 
Let's ship this PR and use another PR to make things more modularized, as it involves multiple files and things not touched in this PR itself.

File: train.py (hunk 6 above)
Comment (wconstab) [reply train.py:121]: I am going to land this PR first.  Then make a new PR to try some refactoring including this suggestion, see if we can iterate to something we like.

File: torchtitan/checkpoint.py (hunk 5 above)
Comment (wconstab) [reply torchtitan/checkpoint.py:162]: i'm moving this discussion to the next PR where I really test out lr_scheduler change for looped schedule.
